### PR TITLE
Add article about BC breaks

### LINKED ISSUE #704: [Article idea] Different kinds of Breaking Changes
@ttoine @eternoendless 

Recently I add a discussion on Twitter where the topic of PrestaShop slow BC cycle was discussed. I thought it would be interesting to explain how hard it is to deal with Breaking Changes, especially the different kind of BC (database change, folder renaming ...).

Would it be a good idea for the PrestaShop Build blog, or would it be too specific (and then I'll write this article on my own blog) ?

### PR DESCRIPTION
**Pull request by matks**
This article explains what exactly are Backward Compatibility Breaks also known as Breaking Changes or BC breaks and how it influences how an open source project like PrestaShop is built. It explains the different types of BC breaks, why they are needed sometimes and how PrestaShop maintainers manage them.

Fixes https://github.com/PrestaShop/prestashop.github.io/issues/704

### CODE REVIEW

**Comment by matks** [comment]
I should improve this article
- quote some usecases of BC breaks done by PrestaShop as examples
- explain it makes it hard for module/theme developers to have big compatibility range
- explain dependency BC break related to compatibility drop and it becomes an issue with release lifecycles accelerating

**Comment by matks** [comment]
I should mention "adding a new argument in __construct()" as a typical BC break

**Comment by matks** [comment]
I need to make sure SQL tables are displayed correctly.

**Comment by matks** [comment]
I should quote some external sources:
- https://en.wikipedia.org/wiki/Backward_compatibility
- https://www.snoyman.com/blog/2018/04/stop-breaking-compatibility
- https://www.snoyman.com/blog/2017/04/enough-with-backwards-compatibility
- https://coenjacobs.me/blog/2013/12/deal-with-breaking-backwards-compatibility/
- https://symfony.com/doc/current/contributing/code/bc.html

**Comment by matks** [comment]
Article is complete from my point of view 😄 I await reviews.

**Review by Progi1984 — APPROVED** [review]
Beautiful

File: news/_posts/2020-06-18-about-bc-breaks.md
Comment (ttoine) [inline news/_posts/2020-06-18-about-bc-breaks.md]:
```suggestion
PrestaShop is a Open Source Project that is being built with the idea that people will use it as a foundation to build their own project. Some will use it as a CMS and extend it with modules and themes. Some will use it as a shop system and customize parts of the e-commerce process. Some will even use it for either its Front-Office or Back-Office but handle the other part in another system, like an ERP or a headless CMS, and manage the communication between the two systems through API calls. So there are "external code" that uses PrestaShop "internal code" (the Core).
```

File: news/_posts/2020-06-18-about-bc-breaks.md
Comment (ttoine) [inline news/_posts/2020-06-18-about-bc-breaks.md:150]:
```suggestion
```

File: news/_posts/2020-06-18-about-bc-breaks.md
Comment (ttoine) [inline news/_posts/2020-06-18-about-bc-breaks.md]:
```suggestion
Code BC breaks are the most common and known BC breaks, and so far this article has only been talking about them.
```

File: news/_posts/2020-06-18-about-bc-breaks.md
Comment (ttoine) [inline news/_posts/2020-06-18-about-bc-breaks.md:175]: here it's ok to keep "we" because it's clear that "we" is the prestashop maintainers

File: news/_posts/2020-06-18-about-bc-breaks.md
Comment (ttoine) [inline news/_posts/2020-06-18-about-bc-breaks.md]:
```suggestion
### How PrestaShop open source project maintainers handle this type of BC break
```

File: news/_posts/2020-06-18-about-bc-breaks.md
Comment (ttoine) [inline news/_posts/2020-06-18-about-bc-breaks.md]:
```suggestion
### How PrestaShop open source project maintainers handle this type of BC break
```

File: news/_posts/2020-06-18-about-bc-breaks.md
Comment (ttoine) [inline news/_posts/2020-06-18-about-bc-breaks.md]:
```suggestion
### How PrestaShop open source project maintainers handle this type of BC break
```

**Review by ttoine — COMMENTED** [review]
great article.

**Comment by matks** [comment]
Applied @ttoine feedbacks.
Waiting for @eternoendless feedback because we all know he is the one who writes the best in English from us all

File: news/_posts/2020-06-18-about-bc-breaks.md
Comment (matthieu-rolland) [inline news/_posts/2020-06-18-about-bc-breaks.md]:
```suggestion
This type of BC break is especially hard to deal with because we have no control on these packages. Sometimes, in order to avoid introducing a dependency BC break, [we decided to maintain a package ourselves](https://github.com/PrestaShop/CsaGuzzleBundle) rather than using the latest version because of the hard BC break it would introduce.
```

**Review by matthieu-rolland — CHANGES_REQUESTED** [review]
A small feedback, your article is very informative (even for me as a core developer) and well written, well done 👍

File: news/_posts/2020-06-18-about-bc-breaks.md
Comment (matthieu-rolland) [inline news/_posts/2020-06-18-about-bc-breaks.md]:
```suggestion
PrestaShop is an Open Source Project that is being built with the idea that people will use it as a foundation to build their own project. Some will use it as a CMS and extend it with modules and themes. Some will use it as a shop system and customize parts of the e-commerce process. Some will even use it for either its Front-Office or Back-Office but handle the other part in another system, like an ERP or a headless CMS, and manage the communication between the two systems through API calls. So there are "external code" that uses PrestaShop "internal code" (the Core).
```

File: news/_posts/2020-06-18-about-bc-breaks.md
Comment (matthieu-rolland) [inline news/_posts/2020-06-18-about-bc-breaks.md]: I notice you wrote "a software" at several places, but software is an uncountable word, for example here you'd write 

```suggestion
Unfortunately... BC breaks are necessary for software to evolve. Not all software changes can be done while maintaining backward compatibility, and BC breaks are a necessary step for a project to evolve in time. And [evolution is necessary](https://build.prestashop.com/news/prestashop-in-2019-and-beyond-part-3-the-future-architecture/) to ensure it continues to fit the needs of its user community, who evolve in time. Software that does not evolve is bound to disappear, replaced by better contributors.
```

Or `A software solution`, `A software program`

This is truly nitpicking :D

File: news/_posts/2020-06-18-about-bc-breaks.md
Comment (matthieu-rolland) [reply news/_posts/2020-06-18-about-bc-breaks.md]: `is **an** Open Source Project`

File: news/_posts/2020-06-18-about-bc-breaks.md
Comment (eternoendless) [inline news/_posts/2020-06-18-about-bc-breaks.md]:
```suggestion
This article explains what exactly are [Backward Compatibility](https://en.wikipedia.org/wiki/Backward_compatibility) Breaks (also known as Breaking Changes or BC) breaks and how it influences the development of an open source project like PrestaShop. It explains the different types of BC breaks, why they are sometimes needed, and how they are managed by PrestaShop maintainers.
```

Comment (eternoendless) [inline news/_posts/2020-06-18-about-bc-breaks.md]: > also known as Breaking Changes or BC

Did you realize we often use the same acronym differently?

- BC = Breaking change
- BC Break = Backwards Compatibility Break

I think this should be explained

File: news/_posts/2020-06-18-about-bc-breaks.md
Comment (eternoendless) [inline news/_posts/2020-06-18-about-bc-breaks.md]:
```suggestion
As soon as BC break is introduced, we start getting lot of angry bug reports on GitHub from people saying "my code was working fine with previous version, and is now broken after I upgraded my shop to the new version!"
```

File: news/_posts/2020-06-18-about-bc-breaks.md
Comment (eternoendless) [inline news/_posts/2020-06-18-about-bc-breaks.md]:
```suggestion
The concept of BC Break is not specific to PrestaShop. This is actually a very common issue for all code that is being used by other software: operating systems like Linux, libraries like Guzzle, tools like cURL, frameworks like Symfony... and Content Management Systems (CMSs) like Drupal or PrestaShop.
```

File: news/_posts/2020-06-18-about-bc-breaks.md
Comment (eternoendless) [inline news/_posts/2020-06-18-about-bc-breaks.md]:
```suggestion
In order to make it easier for the people using these codebases as foundations to handle these changes, [Tom Preston Werner](http://tom.preston-werner.com/) has invented a convention: [SemVer](https://semver.org/). [SemVer](https://semver.org/) aims to help people immediately understand, when seeing a new software version, what to expect regarding its compatibility with current software.
```

Comment (eternoendless) [inline news/_posts/2020-06-18-about-bc-breaks.md]: FYI this has already been explained before https://build.prestashop.com/news/a-more-semantic-versioning-scheme/

**Review by eternoendless — COMMENTED** [review]
Haven't finished reading it yet but I have some comments

File: news/_posts/2020-06-18-about-bc-breaks.md
Comment (matks) [reply news/_posts/2020-06-18-about-bc-breaks.md]: it sound good in my head 😅

File: news/_posts/2020-06-18-about-bc-breaks.md
Comment (matks) [reply news/_posts/2020-06-18-about-bc-breaks.md]: This means I need to introduce the concept "dependency as an API" then explain" SemVer & API contract" before being able to explain this. It's going to make this article longer although it's already quite big. Is it alright ?

File: news/_posts/2020-06-18-about-bc-breaks.md
Comment (matks) [reply news/_posts/2020-06-18-about-bc-breaks.md]: I think it's better to re-explain it for the reader's ease (he does not need to go read something else) but I can refer to the blog post as an external reference.
Wdyt ?
- remove this and redirect to previous blog post
- only refer to previous blog post, keep this section content

**Comment by matks** [comment]
@matthieu-rolland @eternoendless Thank you for proof-reading ! I have applied all your feedbacks already.

Here are the left blocking items:
- I need to explain "BC break" versus "Breaking Change" https://github.com/PrestaShop/prestashop.github.io/pull/730/files#r453559306
- waiting for Pablo answer whether I should define BC break as API changes instead of "code breaking https://github.com/PrestaShop/prestashop.github.io/pull/730#discussion_r453570795 it would be more accurate but a little harder to explain for readers unfamiliar with the topic
- waiting For Pablo answer whether I should remove my explanation about SemVer logic https://github.com/PrestaShop/prestashop.github.io/pull/730#discussion_r453734559

**Comment by matks** [comment]
Pfou 😅 I'm trying to introduce concepts such as "What is a dependency ?" "What is an API interface / a contract ?" "What does it mean for a package" and this article is becoming longer and longer.

I might have to split it in 2.

**Comment by matks** [comment]
Following @eternoendless feedback I have rewritten this article. Part 1 will be coming soon https://github.com/PrestaShop/prestashop.github.io/pull/777

**Pull request closed without merging**
